app/src/main/java/io/gnosis/safe/di/Repositories.kt

```diff
@@ -0,0 +1,23 @@
+package io.gnosis.safe.di
+
+import dagger.Lazy
+import io.gnosis.data.repositories.EnsRepository
+import io.gnosis.data.repositories.SafeRepository
+import pm.gnosis.ethereum.EthereumRepository
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class Repositories
+@Inject constructor(
```

Comment:
Just out of curiosity why do we need this intermediate wrapper?

Reply:
Well, "need" is a strong word. All the dependencies are injected using `dagger.Lazy<T>` which is nice, but our repos at least so far have 0 internal state so there is no performance gain there. Maybe eventually there will be. Also, this way we only need one method in our `ApplicationComponent` which reduces clutter. As, unlike for the legacy app, we don't use contructor injection for repos, since we have a separate module and we don't need dagger in that one just for that. Here we have a `RepositoryModule` and only the `Repositories` class gets provided by the `ApplicationComponent`.

Reply:
Afaik dagger injection is always lazy (even if you do not use the lazy step in between). So classes injected via dagger will only be isntantiated when they are injected. Regarding the methods in the ApplicationComponent, it is just pushing them somewhere else, right? in most of the the repos so far you only use one of these repositories in that class.

Another point (would need to look into to be hones), but if you are not exposing the injected class outside of dependency graph you do not need the method in the ApplicationComponent. So for example in the legacy app we had that view model factory and that was a singleton in the same graph as the repositories and it was responsible for the view models. So actually it was not required to expose the repositories in the application component via methods (for reference https://github.com/gnosis/safe-android/blob/safe_v1/app/src/main/java/pm/gnosis/heimdall/di/components/ApplicationComponent.kt) 

Note on the side: the only reason I requested changes to the PR were the missing unit tests, the rest are more comments or out of pure interest (forgot to mention that 😬 )

Reply:
Ah yes, but that's because you could annotate the constructors of the repositories with @Inject , I think. Like in there: https://github.com/gnosis/safe-android/blob/safe_v1/app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/DefaultEnsRepository.kt

In our case, since the repository classes are all in the :data module, and we are not really pulling dagger into it (as we may want to extract at some point) we just have provider methods in the `RepositoryModule` defined here: https://github.com/gnosis/safe-android/blob/feature/address_input/app/src/main/java/io/gnosis/safe/di/modules/RepositoryModule.kt

I guess in a scenario where we do extract this into a dependency pulled from jitpack, then you could just have a provider method in your app component of the specific repos you want to pull in.

And yeah no worries about requesting changes, this by no means in a mergeable state :) I should have requested review once it was properly finished. But there is a lot of feedback I can already work on anyhow. Thanks!

Reply:
That should be unrelated from the annotations, as these annotations just autogenerate code that has to be handwritten here. It depends highly on where the view models are created. I cannot really check in detail right now, so hard to tell ;)

Reply:
As far as I can tell it should also work with the current setup, as you keep using the ViewModelFactory 🤔

Reply:
I think I will create a task for the next sprint to revise the dagger setup. I think we've mixed things the wrong in some places.

Reply:
Just for tying up the reference: https://github.com/gnosis/safe-android/issues/576